nuxeo-features/nuxeo-platform-comment/nuxeo-platform-comment/src/main/java/org/nuxeo/ecm/platform/comment/impl/AbstractCommentManager.java

```diff
@@ -186,4 +198,81 @@ protected void notifyEvent(CoreSession session, String eventType, DocumentModel
         notifyEvent(session, eventType, commentParent, commentDocumentModel);
     }
 
+    /**
+     * Checks if a document has comments.
+     *
+     * @param session the core session
+     * @param document the document model who's comments are being counted
+     * @return {@code true}if comments were found
+     * @since 11.1
+     */
+    protected boolean hasComments(CoreSession session, DocumentModel document) {
+        // CommentManagerImpl doesn't support it.
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Checks if a document has comments from a particular user.
+     *
+     * @param session the core session
+     * @param document the document model who's comments are being counted
+     * @param author the name of the user who's comments are being counted
+     * @return {@code true} if comments by author were found
+     * @since 11.1
+     */
+    protected boolean hasCommentsByUser(CoreSession session, DocumentModel document, String author) {
+        // CommentManagerImpl doesn't support it.
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Handles autosubscription to comment notifications rules.
+     *
+     * @implSpec The default implementation autosubscribes top level document's author on first comment if
+     *           autosubscription is enabled. This relies on the implementation of the abstract hasComments method.<br/>
+     *           The default implementation autosubscribes comment author on his first comment on top level document if
+     *           autosubscription is enabled. This relies on the implementation of the abstract hasCommentsByUser method
+     * @param session the core session
+     * @param commentDocModel the comment
+     * @param document the document being commented
+     * @since 11.1
+     */
```

Comment:
what if we rework the java doc here:

```
 /**
     * Handles auto-subscription to comment notifications rules.
     *
     * @implSpec If the auto-subscription is enabled {@value NotificationService#AUTOSUBSCRIBE_CONFIG_KEY} then
     *           <ul>
     *           <li>We auto-subscribes to the top level document's author on his first comment:
     *           {@link #hasComments(CoreSession, DocumentModel)}</li>
     *           <li>.....: {@link #hasCommentsByUser(CoreSession, DocumentModel, String)}</li>
     *           </ul>
     * @param session the core session
     * @param commentDocModel the comment document model
     * @param document the document being commented
     * @since 11.1
     */
```

avoid duplication ...

Reply:
/**
     * Handles auto-subscription to comment following notifications rules.
     *
     * @implSpec If the auto-subscription is enabled {@value NotificationService#AUTOSUBSCRIBE_CONFIG_KEY} then
     *           <ul>
     *           <li>Auto-subscribes the top level document's author on the first comment. A comment is the first one if
     *           {@link #hasComments(CoreSession, DocumentModel)} returns {@code false}</li>
     *           <li>Auto-subscribes comment author on his first comment on the top level document. A user's comment is
     *           his first if {@link #hasCommentsByUser(CoreSession, DocumentModel, String)} returns
     *           {@code false}</li>
     *           </ul>
     * @param session the core session
     * @param commentDocModel the comment document model
     * @param document the document being commented
     * @since 11.1
     */

Reply:
Then removing all the javadoc on the protected methods I introduce:
`hasComments `- both signature
`handleNotificationAutoSubscriptions`